File: libs/ts/src/facades/channel.ts
/**
 * A publish-channel address (DESIGN-class-facades §4, `DESIGN-channels.md`): the uniform
 * `{ local, northbound, stream:<name> }` routing target the publish facades resolve on.
 *
 * - `LOCAL` — the local/IPC bus (`messaging().publish`). The default.
 * - `NORTHBOUND` — the northbound/cloud broker (`messaging().publishNorthbound`).
 * - `stream(name)` — the named durable telemetry stream (`getStreams().stream(name).append(...)`
 *   via the {@link StreamSink} seam); **only {@link DataFacade} honors it** —
 *   `events()`/`app()` reject a stream channel (they are low-rate control-plane, not bulk
 *   telemetry).
 *
 * Modeled as a discriminated union (not a bare string enum) because the `stream` variant
 * carries a stream name — mirrors the `UnsScope` interface-plus-namesake-const idiom already
 * used in this library (`uns.ts`) rather than the Java value class's `equals`/`hashCode`
 * boilerplate (TS/vitest compares structurally with `toEqual` out of the box).
 */
import { EdgeCommonsError } from "../errors";

/** The routing kind — the discriminant of the {@link Channel} union. */
export type ChannelKind = "local" | "northbound" | "stream";

/** The local/IPC bus channel. */
export interface LocalChannel {
  readonly kind: "local";
}

/** The northbound/cloud channel. */
export interface NorthboundChannel {
  readonly kind: "northbound";
}

/** The named-durable-stream channel. */
export interface StreamChannel {
  readonly kind: "stream";
  readonly streamName: string;
}

/** A publish-channel address — one of {@link LocalChannel}/{@link NorthboundChannel}/{@link StreamChannel}. */
export type Channel = LocalChannel | NorthboundChannel | StreamChannel;

/**
 * Factory/parse namespace for {@link Channel} (mirrors the `UnsScope` const-namesake idiom):
 * `Channel.LOCAL`, `Channel.NORTHBOUND`, `Channel.stream(name)`, `Channel.fromConfig(value)`.
 */
export const Channel = {
  /** The local/IPC bus channel (the default). */
  LOCAL: { kind: "local" } as LocalChannel,

  /** The northbound/cloud channel. */
  NORTHBOUND: { kind: "northbound" } as NorthboundChannel,

  /**
   * The named-durable-stream channel.
   *
   * @throws EdgeCommonsError (kind `Validation`) when `name` is empty
   */
  stream(name: string): StreamChannel {
    if (!name) {
      throw EdgeCommonsError.validation("stream channel name must be non-empty");
    }
    return { kind: "stream", streamName: name };
  },

  /**
   * Parses a config `publish.channel` string into a {@link Channel} (DESIGN-class-facades §4,
   * Option C). Recognized: `"local"` → {@link Channel.LOCAL}; `"northbound"` →
   * {@link Channel.NORTHBOUND}; `"stream:<name>"` → {@link Channel.stream}. Any
   * other (or null/empty/absent) value yields `undefined` so the caller can fall through to its
   * own default.
   */
  fromConfig(value: string | undefined | null): Channel | undefined {
    if (value === undefined || value === null) {
      return undefined;
    }
    const v = value.trim();
    if (v === "") {
      return undefined;
    }
    const lower = v.toLowerCase();
    if (lower === "local") {
      return Channel.LOCAL;
    }
    if (lower === "northbound") {
      return Channel.NORTHBOUND;
    }
    if (lower.startsWith("stream:")) {
      const name = v.slice("stream:".length);
      return name === "" ? undefined : Channel.stream(name);
    }
    return undefined;
  },

  /** `"local"` / `"northbound"` / `"stream:<name>"` — the config-string form (round-trips {@link fromConfig}). */
  toConfigString(channel: Channel): string {
    switch (channel.kind) {
      case "local":
        return "local";
      case "northbound":
        return "northbound";
      case "stream":
        return `stream:${channel.streamName}`;
    }
  },
};

File: libs/rust/src/facades/channel.rs
//! # Channel — publish-channel routing address
//!
//! **One-liner purpose**: The uniform `{ local, northbound, stream:<name> }` routing target the
//! publish facades resolve on (DESIGN-class-facades §4, `DESIGN-channels.md`), mirroring the Java
//! canonical `com.mbreissi.edgecommons.facades.Channel`.

use crate::error::{EdgeCommonsError, Result};

/// A publish-channel address: the uniform `{ local, northbound, stream:<name> }` routing target.
///
/// - [`Channel::Local`] — the local bus (`messaging().publish`). The default.
/// - [`Channel::Northbound`] — the northbound/cloud broker (`messaging().publish_northbound`).
/// - [`Channel::Stream`] — the named durable telemetry stream
///   (`streams().stream(name).append(...)`); **only [`crate::facades::DataFacade`] honors it** —
///   `events()`/`app()` reject a stream channel (they are low-rate control-plane, not bulk
///   telemetry).
///
/// [`Channel::from_config`] parses the config `publish.channel` string (DESIGN-class-facades §4
/// Option C): `"local"`, `"northbound"`, or `"stream:<name>"`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Channel {
    /// The local bus channel (the default).
    Local,
    /// The northbound/cloud channel.
    Northbound,
    /// The named durable-stream channel (`data()` only).
    Stream(String),
}

impl Channel {
    /// The named-durable-stream channel.
    ///
    /// # Errors
    /// [`EdgeCommonsError::Facade`] when `name` is empty.
    pub fn stream(name: impl Into<String>) -> Result<Channel> {
        let name = name.into();
        if name.is_empty() {
            return Err(EdgeCommonsError::Facade(
                "stream channel name must be non-empty".to_string(),
            ));
        }
        Ok(Channel::Stream(name))
    }

    /// Whether this is the [`Channel::Stream`] variant (the check `events()`/`app()` use to
    /// reject a stream routing override).
    pub const fn is_stream(&self) -> bool {
        matches!(self, Channel::Stream(_))
    }

    /// Parses a config `publish.channel` string into a channel (DESIGN-class-facades §4, Option
    /// C). Recognized: `"local"` → [`Channel::Local`]; `"northbound"` →
    /// [`Channel::Northbound`]; `"stream:<name>"` → [`Channel::Stream`]. Any other
    /// (or empty) value yields `None` so the caller can fall through to its own default.
    pub fn from_config(value: &str) -> Option<Channel> {
        let trimmed = value.trim();
        if trimmed.is_empty() {
            return None;
        }
        let lower = trimmed.to_ascii_lowercase();
        if lower == "local" {
            return Some(Channel::Local);
        }
        if lower == "northbound" {
            return Some(Channel::Northbound);
        }
        if let Some(name) = trimmed.strip_prefix("stream:") {
            return if name.is_empty() {
                None
            } else {
                Some(Channel::Stream(name.to_string()))
            };
        }
        None
    }
}

impl std::fmt::Display for Channel {
    /// `"local"` / `"northbound"` / `"stream:<name>"` — the config-string form.
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Channel::Local => f.write_str("local"),
            Channel::Northbound => f.write_str("northbound"),
            Channel::Stream(name) => write!(f, "stream:{name}"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn from_config_recognizes_all_forms() {
        assert_eq!(Channel::from_config("local"), Some(Channel::Local));
        assert_eq!(
            Channel::from_config("northbound"),
            Some(Channel::Northbound)
        );
        assert_eq!(
            Channel::from_config("stream:hot"),
            Some(Channel::Stream("hot".to_string()))
        );
        assert_eq!(Channel::from_config(""), None);
        assert_eq!(Channel::from_config("iotcore"), None);
        assert_eq!(Channel::from_config("iot_core"), None);
        assert_eq!(Channel::from_config("stream:"), None);
        assert_eq!(Channel::from_config("bogus"), None);
    }

    #[test]
    fn stream_rejects_empty_name() {
        assert!(Channel::stream("").is_err());
        assert!(Channel::stream("hot").is_ok());
    }

    #[test]
    fn is_stream_and_display() {
        assert!(Channel::stream("hot").unwrap().is_stream());
        assert!(!Channel::Local.is_stream());
        assert_eq!(Channel::Northbound.to_string(), "northbound");
        assert_eq!(Channel::Stream("hot".into()).to_string(), "stream:hot");
    }
}
